File: 110-balanced-binary-tree/balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {

    const dfs = (root) => {
        if (!root) {
            return [true,0]
        }
        let left = dfs(root.left)
        let right = dfs(root.right)
        const balanced = (Math.abs(right[1] - left[1]) <= 1) && left[0] && right[0]
        return [balanced, 1 + Math.max(left[1], right[1])]
    }
    return dfs(root)[0]
    
};

File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        

        def dfs(node):
            if node is None:
                return [True, 0]
            
            left_subtree = dfs(node.left)
            right_subtree = dfs(node.right)

            is_balanced = left_subtree[0] and right_subtree[0] and abs(left_subtree[1] - right_subtree[1]) <= 1

            return [is_balanced, max(left_subtree[1], right_subtree[1]) + 1]
        

        return dfs(root)[0]